### ISSUE TITLE
RStudio silently fails to save UTF-8 documents on Windows

### ISSUE THREAD

**Issue by kevinushey** [rstudio/rstudio#2315]
### System details

    RStudio Edition : Desktop
    RStudio Version : v1.2.353
    OS Version      : Windows 10 64bit
    R Version       : R 3.4.3

### Steps to reproduce the problem

With no project open, create a document with the contents:

```
鬼鬼鬼
鬼鬼鬼
鬼鬼鬼
```

Then, attempt to save the document with <kbd>Ctrl + S</kbd>.

### Describe the problem in detail

RStudio seems to indicate that the document was successfully saved, but the file on the filesystem is actually blank.

### Describe the behavior you expected

RStudio should save the file, or at least warn that the save failed because it failed to convert to the requested encoding.

### Notes

This is the RPC RStudio attempts to execute:

<img width="335" alt="screen shot 2018-03-03 at 1 05 44 pm" src="https://user-images.githubusercontent.com/1976582/36939315-a10bb28c-1ee3-11e8-929c-af98b6d815be.png">

With RStudio v1.1, the IDE does warn that the save attempt failed:

> Not all of the characters in ~/scratch/ouch.txt could be encoded using ISO8859-1. To save using a different encoding, choose "File | Save with Encoding..." from the main menu.

And the characters are replaced with `???` markers instead.

**Comment by javierluraschi** [comment]
Here are some findings...

The problem was introduced in build `1.2. 246` (around January 15th), the build `1.2.240` (January 11th) works fine.

The problem is that, somehow, `Riconv` (see ftp://cran.r-project.org/pub/R/doc/manuals/r-patched/R-exts.html#Re_002dencoding), is no longer setting `errno`. This is odd since as the documentation mentions, and since the R-wrapper is a thin wrapper to iconv, when a call fails to `iconv` the `errno` should be set. More specifically, we are hitting this code: 

https://github.com/rstudio/rstudio/blob/9f36698d1a443bfddce69e1eaf1a81c959b6f57a/src/cpp/r/RUtil.cpp#L142-L148

It should never be the case that `result == (size_t)(-1)` is true but `errno` is set to zero, for windows, this should be set to `42` but after build `1.2.246` this is now set to zero.

More interestingly... if we compile an `Rcpp` package with the following code...

```c++
#include <Rcpp.h>
#include <R_ext/Riconv.h>

#include <malloc.h>

using namespace Rcpp;

// [[Rcpp::export]]
List rcpp_hello_world() {
    
    void* handle = ::Riconv_open ("ISO8859-1", "UTF-8");
    
    const char* in = "鬼鬼鬼";
    char * out = (char*)malloc(sizeof(char) * 1024);
    size_t in_size= strlen(in);
    size_t out_size = in_size;
    size_t result = ::Riconv(handle, &in, &in_size, &out, &out_size);
    ::Riconv_close(handle);

    CharacterVector x = CharacterVector::create( "foo", "bar" )  ;
    NumericVector y   = NumericVector::create( (float) result, (float) errno ) ;
    List z            = List::create( x, y ) ;

    return z ;
}
```

it returns `42` for `errno` while calling `rcpp_hello_world()`, but while this exact same code is run from the `rsession` inside RStudios codebase, the `errno` value is set to `0`.

My best guess is that something changed when with the `MSVC` migration and something is being linked incorrectly that we avoid while running this code from the dynamically loaded library through Rcpp.

@kevinushey does this sound related to https://github.com/rstudio/rstudio/blob/master/src/cpp/tools/dll2lib.R?

@gtritchie any interesting bits you can sense from this investigation?

**Comment by gtritchie** [comment]
Well, as you probably know, non-ascii-character handling on Windows is complicated. It prefers to be using 16-bit Unicode via wchar_t / std::wstring, but most of our code assumes utf-8 held in char's. I'm not familiar with how the internals of R deal with this.

MSVC gives quite a few warnings when building RStudio of the "wchar_t being converted to char" or the other way around, so maybe some of those are actually causing us to mishandle wide chars? But not sure if this is a Unicode/wchar_t thing, given the code above that has const char* in = "鬼鬼鬼"; that's either using utf-8 or a Windows code-page?

In other words, no specific ideas at the moment, other than I'm not too surprised we're seeing some issues like this with the switchover...

**Comment by gtritchie** [comment]
BTW, I'd recommend assigning issue to yourself, and setting the "in-progress" label (assuming you're actively working on it). As the team (and number of issues!) grows, we run increasing risk of finding ourselves working on the same things at the same time.

**Comment by dfalty** [comment]
Verified fixed on version 1.2.1090.
